File: packages/dart/lib/src/limits.dart
class ParserLimits {
  const ParserLimits({
    required this.maxVaultBytes,
    required this.maxPayloadBytes,
    required this.maxJsonNesting,
    required this.maxKeyCount,
    required this.maxKeyBytes,
    required this.maxExtensionBytes,
    required this.maxUnlockSlots,
    required this.maxTombstones,
    required this.maxExtensions,
    required this.maxArgon2MemoryKiB,
    required this.maxArgon2Time,
    required this.maxArgon2Parallelism,
    required this.minArgon2MemoryKiB,
    required this.minArgon2Time,
    required this.minArgon2Parallelism,
    required this.maxIdentityBytes,
  });

  final int maxVaultBytes;
  final int maxPayloadBytes;
  final int maxJsonNesting;
  final int maxKeyCount;
  final int maxKeyBytes;
  final int maxExtensionBytes;
  final int maxUnlockSlots;
  final int maxTombstones;
  final int maxExtensions;
  final int maxArgon2MemoryKiB;
  final int maxArgon2Time;
  final int maxArgon2Parallelism;
  final int minArgon2MemoryKiB;
  final int minArgon2Time;
  final int minArgon2Parallelism;
  final int maxIdentityBytes;
}

const defaultLimits = ParserLimits(
  maxVaultBytes: 16 * 1024 * 1024,
  maxPayloadBytes: 16 * 1024 * 1024,
  maxJsonNesting: 32,
  maxKeyCount: 1024,
  maxKeyBytes: 1024 * 1024,
  maxExtensionBytes: 64 * 1024,
  maxUnlockSlots: 64,
  maxTombstones: 1024,
  maxExtensions: 64,
  maxArgon2MemoryKiB: 1048576,
  maxArgon2Time: 16,
  maxArgon2Parallelism: 16,
  minArgon2MemoryKiB: 16384,
  minArgon2Time: 2,
  minArgon2Parallelism: 1,
  maxIdentityBytes: 2048,
);

class Argon2idParams {
  const Argon2idParams({
    this.alg = 'Argon2id',
    required this.m,
    required this.t,
    required this.p,
    this.keyLength = 32,
  });

  final String alg;
  final int m;
  final int t;
  final int p;
  final int keyLength;
}

const recommendedArgon2id = Argon2idParams(m: 65536, t: 3, p: 4);

/// Legal for tests/CI; still meets SPEC §6.5 minima.
const testArgon2id = Argon2idParams(m: 16384, t: 2, p: 1);

File: packages/js/packages/core/src/limits.ts
export interface ParserLimits {
  maxVaultBytes: number;
  maxPayloadBytes: number;
  maxJsonNesting: number;
  maxKeyCount: number;
  maxKeyBytes: number;
  maxExtensionBytes: number;
  maxUnlockSlots: number;
  maxTombstones: number;
  maxExtensions: number;
  maxArgon2MemoryKiB: number;
  maxArgon2Time: number;
  maxArgon2Parallelism: number;
  minArgon2MemoryKiB: number;
  minArgon2Time: number;
  minArgon2Parallelism: number;
  maxIdentityBytes: number;
}

/** SPEC Appendix C defaults. Applications MAY tighten; they MUST NOT silently raise KDF mins. */
export const DEFAULT_LIMITS: ParserLimits = {
  maxVaultBytes: 16 * 1024 * 1024,
  maxPayloadBytes: 16 * 1024 * 1024,
  maxJsonNesting: 32,
  maxKeyCount: 1024,
  maxKeyBytes: 1024 * 1024,
  maxExtensionBytes: 64 * 1024,
  maxUnlockSlots: 64,
  maxTombstones: 1024,
  maxExtensions: 64,
  maxArgon2MemoryKiB: 1_048_576,
  maxArgon2Time: 16,
  maxArgon2Parallelism: 16,
  minArgon2MemoryKiB: 16_384,
  minArgon2Time: 2,
  minArgon2Parallelism: 1,
  maxIdentityBytes: 2048,
};

export const RECOMMENDED_ARGON2ID = {
  alg: "Argon2id" as const,
  m: 65_536,
  t: 3,
  p: 4,
  key_length: 32,
};

/** Legal for tests/CI; still meets SPEC §6.5 minima. */
export const TEST_ARGON2ID = {
  alg: "Argon2id" as const,
  m: 16_384,
  t: 2,
  p: 1,
  key_length: 32,
};
